src/util.c

```diff
@@ -18,17 +18,21 @@ writable_buffer_ensure_space(writable_buffer *wb, uint64 minspace)
       minspace = 2 * wb->buffer_capacity;
    }
 
-   void *oldptr  = wb->can_free ? wb->buffer : NULL;
-   void *newdata = platform_realloc(
-      wb->heap_id, writable_buffer_length(wb), oldptr, minspace);
+   void *newdata = NULL;
+   if (wb->can_free) {
```

Comment:
@rtjohnso - Reworked this logic as discussed. Newly added unit-tests seem to work cleanly.